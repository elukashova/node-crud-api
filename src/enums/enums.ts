export enum Requests {
  Get = 'GET',
  Post = 'POST',
  Put = 'PUT',
  Delete = 'DELETE',
}

export enum Actions {
  GetUsers,
  GetUser,
  CreateUser,
  UpdateUser,
  DeleteUser,
}

export enum Endpoints {
  Base = '/api/users',
  GetUsers = '/api/users',
  GetUser = '/api/users/',
  CreateUser = '/api/users',
  UpdateUser = '/api/users/',
  DeleteUser = '/api/users/',
}

export enum StatusCodes {
  Ok = 200,
  Created = 201,
  NoContent = 204,
  BadRequest = 400,
  NotFound = 404,
  InternalServerError = 500,
}

export enum Errors {
  Message404NotFound = 'The requested resource cannot be found by the server.',
  Message404User = 'The user with this id cannot be found by the server.',
  Message404Endpoint = 'Ooops, this endpoint doesn\'t seem correct. Please retry.',
  Message400Uuid = 'Invalid UUID. Please, provide a correct one.',
  Message400Body = 'Invalid user data. Please, provide user\'s \'username\', \'age\' and [hobbies].',
  Message500Server = 'The server encountered an error and could not complete your request.'
}