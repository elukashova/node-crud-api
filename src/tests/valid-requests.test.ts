import 'dotenv/config';
import supertest from 'supertest';
import { StatusCodes, Endpoints, Errors } from '../enums/enums';

const port = +process.env.PORT! || 4000; 
const request = supertest(`http://localhost:${port}`); 

describe('CRUD API', () => {
    test('should get all users', () => {
        request
            .get(Endpoints.GetUsers)
            .end((err, res) => {
                expect(res.status).toBe(StatusCodes.Ok);
                expect(Array.isArray(JSON.parse(res.text))).toBe(true);
                expect(JSON.parse(res.text)).toHaveLength(0);
            })
    });

    let userID: string;
    const dummyUser = {
        username: 'test',
        age: 10,
        hobbies: ['test', 'test'],
    };

    test('should create a new user', done => {
        request
            .post(Endpoints.CreateUser)
            .send(JSON.stringify(dummyUser))
            .end((err, res) => {
                expect(res.status).toBe(StatusCodes.Created);
                expect(JSON.parse(res.text)).toHaveProperty('id');
                expect(JSON.parse(res.text)).toMatchSnapshot({id: expect.any(String), username: dummyUser.username, age: dummyUser.age, hobbies: dummyUser.hobbies});
                userID = JSON.parse(res.text).id;
                done();
            });
    });

    test('should get a user by id', () => {
        request
            .get(`${Endpoints.GetUser}${userID}`)
            .end((err, res) => {
                expect(res.status).toBe(StatusCodes.Ok);
                expect(JSON.parse(res.text)).toHaveProperty('id');
                expect(JSON.parse(res.text).id).toMatch(userID);
                expect(JSON.parse(res.text)).toHaveProperty('username');
                expect(JSON.parse(res.text)).toHaveProperty('age');
                expect(JSON.parse(res.text)).toHaveProperty('hobbies');
            });
    });

    const updatedDummyUser = {
        username: 'test2',
        age: 12,
        hobbies: ['test2', 'test2'],
    };

    test('should update user data', () => {
        request
            .put(`${Endpoints.UpdateUser}${userID}`)
            .send(JSON.stringify(updatedDummyUser))
            .end((err, res) => {
                expect(JSON.parse(res.text)).toMatchSnapshot({id: userID, username: 'test2', age: 12, hobbies: ['test2', 'test3']});
            });
    });

    test('should delete user data', () => {
        request
            .delete(`${Endpoints.DeleteUser}${userID}`)
            .end((err, res) => {
                expect(res.status).toBe(StatusCodes.NoContent);
            });
    });
});
