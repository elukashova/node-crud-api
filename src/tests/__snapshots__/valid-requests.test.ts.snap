// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CRUD API should create a new user 1`] = `
{
  "age": 10,
  "hobbies": [
    "test",
    "test",
  ],
  "id": Any<String>,
  "username": "test",
}
`;
